python_modules/dagster/dagster/_core/execution/context/compute.py

```diff
@@ -1223,8 +1245,461 @@ def asset_check_spec(self) -> AssetCheckSpec:
+    "get_tag": "use dagster_run.get_tag instead",
+    "run_tags": "use dagster_run.tags instead",
+    "set_data_version": "use MaterializeResult instead",
+    "run": "use dagster_run instead.",
+}
+
+# TODO - add AssetCheck related methods to this list
+
+
+def _get_deprecation_kwargs(attr: str):
+    deprecation_kwargs = {"breaking_version": "1.7.0"}
+    deprecation_kwargs["subject"] = f"AssetExecutionContext.{attr}"
+
+    if attr in OP_EXECUTION_CONTEXT_ONLY_METHODS:
+        deprecation_kwargs["additional_warn_text"] = (
+            f"You have called the deprecated method {attr} on AssetExecutionContext. Use"
+            " the underlying OpExecutionContext instead by calling"
+            f" context.op_execution_context.{attr}."
+        )
+
+    if attr in DEPRECATED_IO_MANAGER_CENTRIC_CONTEXT_METHODS:
+        alt = DEPRECATED_IO_MANAGER_CENTRIC_CONTEXT_METHODS[attr]
+        deprecation_kwargs["additional_warn_text"] = (
+            f"You have called method {attr} on AssetExecutionContext that is oriented"
+            f" around I/O managers. If you not using I/O managers we suggest you {alt}. If"
+            " you are using I/O managers the method still exists at"
+            f" context.op_execution_context.{attr}."
+        )
+
+    if attr in ALTERNATE_AVAILABLE_METHODS:
+        deprecation_kwargs["additional_warn_text"] = f"Instead {ALTERNATE_AVAILABLE_METHODS[attr]}."
+
+    return deprecation_kwargs
+
+
+class AssetExecutionContext(OpExecutionContext):
+    def __init__(self, op_execution_context: OpExecutionContext) -> None:
+        self._op_execution_context = check.inst_param(
+            op_execution_context, "op_execution_context", OpExecutionContext
+        )
+
+    @public
+    @property
+    def op_execution_context(self) -> OpExecutionContext:
+        return self._op_execution_context
+
+    # IContext interface methods
+
+    @property
+    def is_asset_step(self) -> bool:
+        return self.op_execution_context.has_assets_def
+
+    @public
+    @property
+    def asset_key(self) -> AssetKey:
+        return self._op_execution_context.asset_key
+
+    @property
+    def asset_keys(self) -> Sequence[AssetKey]:
+        return list(self.op_execution_context.assets_def.keys_by_output_name.values())
+
+    @property
+    def provenance(self) -> Optional[DataProvenance]:
+        return self.get_asset_provenance(self.asset_key)
+
+    @property
+    def provenance_by_asset_key(self) -> Mapping[AssetKey, Optional[DataProvenance]]:
+        provenance_map = {}
+        for key in self.asset_keys:
+            provenance_map[key] = self.get_asset_provenance(key)
+
+        return provenance_map
+
+    @property
+    def code_version(self) -> Optional[str]:
+        return self.get_assets_code_version([self.asset_key])[self.asset_key]
+
+    @property
+    def code_version_by_asset_key(self) -> Mapping[AssetKey, Optional[str]]:
+        return self.get_assets_code_version(self.asset_keys)
```

Comment:
these all go to the db? Super dangerous to make that this easy

Reply:
resolved via call - code version is now fetched from the asset definition, and provenance for the selected assets is pre-fetched during init